### ISSUE TITLE
Implementations of ASTRule are quite repetitive

### ISSUE THREAD

**Issue by aarondaub** [realm/SwiftLint#73]
I'm not sure what an appropriate solution to this would be but ASTRule implementations have a fair bit of code overlap.

```swift
  func validateFile(file: File) 
```

just calls

```swift
  func validateFile(file: File, dictionary: XPCDictionary)
```

which just calls

```swift
  func validateFile(file:File, kind: SwiftDeclarationKind, dictionary: XPCDictionary)
```

repeatedly for some substructures.

**Comment by aarondaub** [comment]
So it looks like 

```swift
  func validateFile(file: File, dictionary: XPCDictionary)
```

So far every implementation does the same thing, and I can't envision a reason for it to differ for specific rules. We could make a free function that does the enumeration (it would have to take an ASTRule as well). Thoughts?

**Comment by jpsim** [comment]
Yes, those rules have quite a bit of duplicated code. Please feel free to simplify them or make them share more code.

**Comment by jpsim** [comment]
Fixed in #186.
